diff-hl.el

```diff
@@ -1525,42 +1534,119 @@ view buffer.
+         (diff-hl-reset-reference-rev '(4)))
+        (t (setq-default diff-hl-reference-revision rev)
+           (dolist (buf (buffer-list))
+             (with-current-buffer buf
+               (when (and diff-hl-mode
+                          (not (local-variable-p 'diff-hl-reference-revision)))
+                 (when (bound-and-true-p diff-hl-amend-mode)
+                   (diff-hl-amend-mode -1))
+                 (diff-hl-update)))))))
 
 ;;;###autoload
-(defun diff-hl-reset-reference-rev ()
-  "Reset the reference revision globally to the most recent one."
-  (interactive)
-  (setq diff-hl-reference-revision nil)
+(defun diff-hl-set-reference-rev-in-project (rev)
+  "Set the reference revision in the current project to REV.
+When called interactively, REV read with completion.
+
+When called with a prefix argument, reset to the most recent one instead.
+
+The default value chosen using one of methods below:
+
+- In a log view buffer, it uses the revision of current entry.
+Call `vc-print-log' or `vc-print-root-log' first to open a log
+view buffer.
+- In a VC annotate buffer, it uses the revision of current line.
+- In other situations, it uses the symbol at point.
+
+Projects whose reference was set with this command are unaffected by
+subsequent changes to the global reference (see
+`diff-hl-set-reference-rev`).
+
+Also notice that this will disable `diff-hl-amend-mode' in
+buffers that enables it, since `diff-hl-amend-mode' overrides its
+effect."
+  (interactive
+   (if current-prefix-arg current-prefix-arg
+     (let* ((def (or (and (equal major-mode 'vc-annotate-mode)
+                          (car (vc-annotate-extract-revision-at-line)))
+                     (log-view-current-tag)
+                     (thing-at-point 'symbol t)))
+            (prompt (if def
+                        (format "Reference revision (default %s): " def)
+                      "Reference revision: ")))
+       (list (vc-read-revision prompt nil nil def)))))
+  (unless rev
+    (user-error "No reference revision specified"))
+  (let ((name (project-name (project-current))))
+    (cond
+     ;; reset
+     (current-prefix-arg
+      (diff-hl-set-reference-rev-in-project-internal nil)
+      (message "Reference revision reset to the most recent revision (project %s)" name))
+     ;; set
+     (t (diff-hl-set-reference-rev-in-project-internal rev)
+        (message "Showing changes against %s (project %s)" rev name)))))
+
+(defun diff-hl-set-reference-rev-in-project-internal (rev)
+  (let* ((proj (project-current))
+         (class (intern (concat "diff-hl-" (project-name proj)))))
+    ;; newly opened files will share this value
+    (dir-locals-set-class-variables class
+                                    `((nil (diff-hl-reference-revision . ,rev))))
+    ;; run once per project
+    (unless (assq class dir-locals-directory-cache)
+      (dir-locals-set-directory-class (project-root proj)
+                                      class))
```

Comment:
@dgutov shouldn't this be something project.el handles, if we want to have project-local variables implemented this way?

Reply:
Good question.

@EugeneNeuron Do you think that instead we would want to use `add-dir-local-variable`, to alter (or create) a corresponding `.dir-locals.el` file, adding such a variable there?

This can have a benefit that a) it persists between sessions, b) you can fairly easily find out which value is used for which project, instead of trying introspection on the values altered by `dir-locals-set-directory-class`.